root/static/scripts/common/components/TagEditor.js

```diff
@@ -362,32 +357,41 @@ class TagEditor extends React.Component<TagEditorProps, TagEditorState> {
 
   updateTags(updatedUserTags: $ReadOnlyArray<TagUpdateT>) {
     const newTags = this.state.tags.slice(0);
-
     updatedUserTags.forEach(t => {
-      const index = _.findIndex(newTags, ct => ct.tag === t.tag);
+      const $ = require('jquery');
+      const index = _.findIndex(newTags, ct => ct.tag.name === t.tag);
+      const url = `/ws/js/genre?q=${encodeURIComponent(t.tag)}&limit=100`;
 
-      if (t.deleted) {
-        newTags.splice(index, 1);
-      } else {
-        const tag = {
-          count: t.count,
-          tag: t.tag,
-          vote: t.vote,
-        };
+      $.getJSON(url, (data) => {
+        const genre = data.find(genre => (genre.name === t.tag));
 
-        if (index >= 0) {
-          newTags[index] = tag;
+        if (t.deleted) {
+          newTags.splice(index, 1);
         } else {
-          newTags.push(tag);
+          const tag = {
+            count: t.count,
+            tag: {
+              entityType: 'tag',
+              genre: genre,
+              id: null,
+              name: t.tag,
+            },
+            vote: t.vote,
+          };
+
+          if (index >= 0) {
+            newTags[index] = tag;
+          } else {
+            newTags.push(tag);
+          }
+          this.setState({tags: sortedTags(newTags)});
```

Comment:
These changes look a bit off since it only needs to make a request if the tag wasn't deleted, and it should only call `setState` once at the end. (It doesn't appear to call `setState` at all for deleted tags. It's also not impossible that a callback for one `$.getJSON` call might run before a previous one, or not complete at all.)

But this is kind of tricky to do cleanly, and given that the autocomplete is also made worse by having to use `/ws/js`, I think we should maybe just seed the `TagEditor` with a list of all genre names in the props and mostly stick to the previous implementation. :/ What do you think?

Reply:
We were talking about making it so that there was a way to ask the WS for all genre strings (and deal with https://tickets.metabrainz.org/browse/MBS-9880 at the same time). Not sure what's the best endpoint for that.

Reply:
We should probably have a browse endpoint in /ws/2 that includes the gid, comment, etc. Not certain if we'll have a /ws/2 endpoint that just returns a giant list of names, though, which is basically what we need here. I think we could just add the latter to /ws/js for now (with the list cached in redis).

Reply:
But I think we'll still want to pass the list as a prop, rather than fetching them from /ws/js in the component, because the component should still display something if the user isn't logged in or has JavaScript disabled.

Reply:
Updated this (passing it through the stash since it seems easier that way than having it as a prop on every place that has genres on the sidebar). Quick test seems to work, but it can certainly use more eyes.